File: dynagox_flink/src/main/java/dynagox/experiments/ceb/DynaGoxCEB3.java
package dynagox.experiments.ceb;

import dynagox.experiments.Utils;
import dynagox.experiments.hashmap.LinkedListNode;
import dynagox.experiments.hashmap.MultiHashMap;
import dynagox.experiments.hashmap.PrimaryHashIndex;
import dynagox.experiments.hashmap.SecondaryHashIndex;
import dynagox.experiments.common.Tuple1;
import dynagox.experiments.common.Tuple2;

import java.util.List;

public class DynaGoxCEB3 extends DynaGoxCEB {
    public DynaGoxCEB3(double epsilon, boolean print) {
        apv_an = new MultiHashMap<>(new PrimaryHashIndex<>(new Tuple1.Hash(), new Tuple1.Equal()), List.of());

        r_ci = new MultiHashMap<>(
                new PrimaryHashIndex<>(new Tuple2.Hash(), new Tuple2.Equal()),
                List.of(new SecondaryHashIndex<>(new Tuple2.Hash1(), new Tuple2.Equal1())));

        apv_ci = new MultiHashMap<>(new PrimaryHashIndex<>(new Tuple1.Hash(), new Tuple1.Equal()), List.of());

        apv_mi = new MultiHashMap<>(new PrimaryHashIndex<>(new Tuple1.Hash(), new Tuple1.Equal()), List.of());

        apv_mii1 = new MultiHashMap<>(new PrimaryHashIndex<>(new Tuple1.Hash(), new Tuple1.Equal()), List.of());

        apv_mii2 = new MultiHashMap<>(new PrimaryHashIndex<>(new Tuple1.Hash(), new Tuple1.Equal()), List.of());

        apv_pi = new MultiHashMap<>(new PrimaryHashIndex<>(new Tuple1.Hash(), new Tuple1.Equal()), List.of());

        k_an = new Tuple1();
        t_ci = new Tuple2();
        k_ci = new Tuple1();
        k_mi = new Tuple1();
        k_mii1 = new Tuple1();
        k_mii2 = new Tuple1();
        k_pi = new Tuple1();

        b = Utils.get_b(epsilon, 1);
        print_result = print;
    }

    @Override
    public void insert_into_aka_name(int id, int personId) {
        k_an.f0 = personId;
        apv_an.update_add(k_an, 1);
        k_pi.f0 = personId;
        Tuple1 g_pi = apv_pi.get(k_pi);
        if (g_pi != null) {
            k_ci.f0 = personId;
            Tuple1 g_ci = apv_ci.get(k_ci);
            if (g_ci != null) {
                result += g_ci.get_annotation() * g_pi.get_annotation();
            }
        }
    }

    @Override
    public void insert_into_cast_info(int personId, int movieId) {
        t_ci.f0 = personId;
        t_ci.f1 = movieId;
        r_ci.update_add(t_ci, 1);
        k_mi.f0 = movieId;
        Tuple1 g_mi = apv_mi.get(k_mi);
        if (g_mi != null) {
            k_mii1.f0 = movieId;
            Tuple1 g_mii1 = apv_mii1.get(k_mii1);
            if (g_mii1 != null) {
                k_mii2.f0 = movieId;
                Tuple1 g_mii2 = apv_mii2.get(k_mii2);
                if (g_mii2 != null) {
                    k_ci.f0 = personId;
                    apv_ci.update_add(k_ci, g_mi.get_approximate() * g_mii1.get_annotation() * g_mii2.get_annotation());
                    k_an.f0 = personId;
                    Tuple1 g_an = apv_an.get(k_an);
                    if (g_an != null) {
                        k_pi.f0 = personId;
                        Tuple1 g_pi = apv_pi.get(k_pi);
                        if (g_pi != null) {
                            result += g_an.get_annotation() * g_mi.get_approximate() * g_mii1.get_annotation() * g_mii2.get_annotation() * g_pi.get_annotation();
                        }
                    }
                }
            }
        }
    }

    @Override
    public void insert_into_movie_info(int id, int movieId) {
        k_mi.f0 = movieId;
        double d_mi = apv_mi.update_add_approx(k_mi, 1, b);
        if (d_mi != 0) {
            k_mii1.f0 = movieId;
            Tuple1 g_mii1 = apv_mii1.get(k_mii1);
            if (g_mii1 != null) {
                k_mii2.f0 = movieId;
                Tuple1 g_mii2 = apv_mii2.get(k_mii2);
                if (g_mii2 != null) {
                    t_ci.f1 = movieId;
                    LinkedListNode<Tuple2> f_ci = r_ci.find(0, t_ci);
                    while (f_ci != null) {
                        k_ci.f0 = f_ci.object.f0;
                        apv_ci.update_add(k_ci, f_ci.object.get_annotation() * d_mi * g_mii1.get_annotation() * g_mii2.get_annotation());
                        k_an.f0 = f_ci.object.f0;
                        Tuple1 g_an = apv_an.get(k_an);
                        if (g_an != null) {
                            k_pi.f0 = f_ci.object.f0;
                            Tuple1 g_pi = apv_pi.get(k_pi);
                            if (g_pi != null) {
                                result += g_an.get_annotation() * f_ci.object.get_annotation() * d_mi * g_mii1.get_annotation() * g_mii2.get_annotation() * g_pi.get_annotation();
                            }
                        }
                        f_ci = f_ci.next;
                    }
                }
            }
        }
    }

    @Override
    public void insert_into_movie_info_idx1(int id, int movieId) {
        k_mii1.f0 = movieId;
        apv_mii1.update_add(k_mii1, 1);
        k_mi.f0 = movieId;
        Tuple1 g_mi = apv_mi.get(k_mi);
        if (g_mi != null) {
            k_mii2.f0 = movieId;
            Tuple1 g_mii2 = apv_mii2.get(k_mii2);
            if (g_mii2 != null) {
                t_ci.f1 = movieId;
                LinkedListNode<Tuple2> f_ci = r_ci.find(0, t_ci);
                while (f_ci != null) {
                    k_ci.f0 = f_ci.object.f0;
                    apv_ci.update_add(k_ci, f_ci.object.get_annotation() * g_mi.get_approximate() * g_mii2.get_annotation());
                    k_an.f0 = f_ci.object.f0;
                    Tuple1 g_an = apv_an.get(k_an);
                    if (g_an != null) {
                        k_pi.f0 = f_ci.object.f0;
                        Tuple1 g_pi = apv_pi.get(k_pi);
                        if (g_pi != null) {
                            result += g_an.get_annotation() * f_ci.object.get_annotation() * g_mi.get_approximate() * g_mii2.get_annotation() * g_pi.get_annotation();
                        }
                    }
                    f_ci = f_ci.next;
                }
            }
        }
    }

    @Override
    public void insert_into_movie_info_idx2(int id, int movieId) {
        k_mii2.f0 = movieId;
        apv_mii2.update_add(k_mii2, 1);
        k_mi.f0 = movieId;
        Tuple1 g_mi = apv_mi.get(k_mi);
        if (g_mi != null) {
            k_mii1.f0 = movieId;
            Tuple1 g_mii1 = apv_mii1.get(k_mii1);
            if (g_mii1 != null) {
                t_ci.f1 = movieId;
                LinkedListNode<Tuple2> f_ci = r_ci.find(0, t_ci);
                while (f_ci != null) {
                    k_ci.f0 = f_ci.object.f0;
                    apv_ci.update_add(k_ci, f_ci.object.get_annotation() * g_mi.get_approximate() * g_mii1.get_annotation());
                    k_an.f0 = f_ci.object.f0;
                    Tuple1 g_an = apv_an.get(k_an);
                    if (g_an != null) {
                        k_pi.f0 = f_ci.object.f0;
                        Tuple1 g_pi = apv_pi.get(k_pi);
                        if (g_pi != null) {
                            result += g_an.get_annotation() * f_ci.object.get_annotation() * g_mi.get_approximate() * g_mii1.get_annotation() * g_pi.get_annotation();
                        }
                    }
                    f_ci = f_ci.next;
                }
            }
        }
    }

    @Override
    public void insert_into_person_info(int id, int personId) {
        k_pi.f0 = personId;
        apv_pi.update_add(k_pi, 1);
        k_an.f0 = personId;
        Tuple1 g_an = apv_an.get(k_an);
        if (g_an != null) {
            k_ci.f0 = personId;
            Tuple1 g_ci = apv_ci.get(k_ci);
            if (g_ci != null) {
                result += g_an.get_annotation() * g_ci.get_annotation();
            }
        }
    }
}

File: dynagox_cpp/implement/experiments/src/dynagox/dynagox_ceb3.cpp
#include "dynagox_ceb3.h"

void DynaGoxCEB3::process(const std::string &line) {
  std::stringstream stream(line);
  std::string field;

  getline(stream, field, '|');
  unsigned long relation_id = std::stoul(field);
  if (relation_id == 0) {
    getline(stream, field, '|');
    unsigned long id = std::stoul(field);
    getline(stream, field, '|');
    unsigned long person_id = std::stoul(field);
    insert_into_an(id, person_id);
  } else if (relation_id == 1) {
    getline(stream, field, '|');
    unsigned long id = std::stoul(field);
    getline(stream, field, '|');
    unsigned long person_id = std::stoul(field);
    getline(stream, field, '|');
    unsigned long movie_id = std::stoul(field);
    insert_into_ci(person_id, movie_id);
  } else if (relation_id == 3) {
    getline(stream, field, '|');
    unsigned long id = std::stoul(field);
    getline(stream, field, '|');
    unsigned long movie_id = std::stoul(field);
    insert_into_mi(id, movie_id);
  } else if (relation_id == 4) {
    getline(stream, field, '|');
    unsigned long id = std::stoul(field);
    getline(stream, field, '|');
    unsigned long movie_id = std::stoul(field);
    insert_into_mii1(id, movie_id);
    insert_into_mii2(id, movie_id);
  } else if (relation_id == 6) {
    getline(stream, field, '|');
    unsigned long id = std::stoul(field);
    getline(stream, field, '|');
    unsigned long person_id = std::stoul(field);
    insert_into_pi(id, person_id);
  } else {
    fmt::println("Unknown relation ID '{}'.", relation_id);
    exit(1);
  }
}

void DynaGoxCEB3::milestone() { print(); }

void DynaGoxCEB3::insert_into_an(unsigned long id, unsigned long person_id) {
  k_an.f0 = person_id;
  apv_an.update_add(k_an, 1);
  k_ci.f0 = person_id;
  auto g_ci = apv_ci.get(k_ci);
  if (g_ci != nullptr) {
    k_pi.f0 = person_id;
    auto g_pi = apv_pi.get(k_pi);
    if (g_pi != nullptr) {
      result += g_ci->annotation * g_pi->annotation;
    }
  }
}

void DynaGoxCEB3::insert_into_ci(unsigned long person_id, unsigned long movie_id) {
  t_ci.f0 = person_id;
  t_ci.f1 = movie_id;
  r_ci.update_add(t_ci, 1);
  k_mi.f0 = movie_id;
  auto g_mi = apv_mi.get(k_mi);
  if (g_mi != nullptr) {
    k_mii1.f0 = movie_id;
    auto g_mii1 = apv_mii1.get(k_mii1);
    if (g_mii1 != nullptr) {
      k_mii2.f0 = movie_id;
      auto g_mii2 = apv_mii2.get(k_mii2);
      if (g_mii2 != nullptr) {
        k_ci.f0 = person_id;
        apv_ci.update_add(k_ci, g_mi->approximate * g_mii1->annotation * g_mii2->annotation);
        k_an.f0 = person_id;
        auto g_an = apv_an.get(k_an);
        if (g_an != nullptr) {
          k_pi.f0 = person_id;
          auto g_pi = apv_pi.get(k_pi);
          if (g_pi != nullptr) {
            result +=
                g_an->annotation * g_mi->approximate * g_mii1->annotation * g_mii2->annotation * g_pi->annotation;
          }
        }
      }
    }
  }
}

void DynaGoxCEB3::insert_into_mi(unsigned long id, unsigned long movie_id) {
  k_mi.f0 = movie_id;
  auto d_mi = apv_mi.update_add_approx(k_mi, 1, b);
  if (d_mi != 0) {
    k_mii1.f0 = movie_id;
    auto g_mii1 = apv_mii1.get(k_mii1);
    if (g_mii1 != nullptr) {
      k_mii2.f0 = movie_id;
      auto g_mii2 = apv_mii2.get(k_mii2);
      if (g_mii2 != nullptr) {
        t_ci.f1 = movie_id;
        auto f_ci = r_ci.find(0, t_ci);
        while (f_ci != nullptr) {
          k_ci.f0 = f_ci->object->f0;
          apv_ci.update_add(k_ci, f_ci->object->annotation * d_mi * g_mii1->annotation * g_mii2->annotation);
          k_an.f0 = f_ci->object->f0;
          auto g_an = apv_an.get(k_an);
          if (g_an != nullptr) {
            k_pi.f0 = f_ci->object->f0;
            auto g_pi = apv_pi.get(k_pi);
            if (g_pi != nullptr) {
              result += g_an->annotation * f_ci->object->annotation * d_mi * g_mii1->annotation * g_mii2->annotation *
                        g_pi->annotation;
            }
          }
          f_ci = f_ci->next;
        }
      }
    }
  }
}

void DynaGoxCEB3::insert_into_mii1(unsigned long id, unsigned long movie_id) {
  k_mii1.f0 = movie_id;
  apv_mii1.update_add(k_mii1, 1);
  k_mi.f0 = movie_id;
  auto g_mi = apv_mi.get(k_mi);
  if (g_mi != nullptr) {
    k_mii2.f0 = movie_id;
    auto g_mii2 = apv_mii2.get(k_mii2);
    if (g_mii2 != nullptr) {
      t_ci.f1 = movie_id;
      auto f_ci = r_ci.find(0, t_ci);
      while (f_ci != nullptr) {
        k_ci.f0 = f_ci->object->f0;
        apv_ci.update_add(k_ci, f_ci->object->annotation * g_mi->approximate * g_mii2->annotation);
        k_an.f0 = f_ci->object->f0;
        auto g_an = apv_an.get(k_an);
        if (g_an != nullptr) {
          k_pi.f0 = f_ci->object->f0;
          auto g_pi = apv_pi.get(k_pi);
          if (g_pi != nullptr) {
            result += g_an->annotation * f_ci->object->annotation * g_mi->approximate * g_mii2->annotation * g_pi->annotation;
          }
        }
        f_ci = f_ci->next;
      }
    }
  }
}

void DynaGoxCEB3::insert_into_mii2(unsigned long id, unsigned long movie_id) {
  k_mii2.f0 = movie_id;
  apv_mii2.update_add(k_mii2, 1);
  k_mi.f0 = movie_id;
  auto g_mi = apv_mi.get(k_mi);
  if (g_mi != nullptr) {
    k_mii1.f0 = movie_id;
    auto g_mii1 = apv_mii1.get(k_mii1);
    if (g_mii1 != nullptr) {
      t_ci.f1 = movie_id;
      auto f_ci = r_ci.find(0, t_ci);
      while (f_ci != nullptr) {
        k_ci.f0 = f_ci->object->f0;
        apv_ci.update_add(k_ci, f_ci->object->annotation * g_mi->approximate * g_mii1->annotation);
        k_an.f0 = f_ci->object->f0;
        auto g_an = apv_an.get(k_an);
        if (g_an != nullptr) {
          k_pi.f0 = f_ci->object->f0;
          auto g_pi = apv_pi.get(k_pi);
          if (g_pi != nullptr) {
            result += g_an->annotation * f_ci->object->annotation * g_mi->approximate * g_mii1->annotation * g_pi->annotation;
          }
        }
        f_ci = f_ci->next;
      }
    }
  }
}

void DynaGoxCEB3::insert_into_pi(unsigned long id, unsigned long person_id) {
  k_pi.f0 = person_id;
  apv_pi.update_add(k_pi, 1);
  k_an.f0 = person_id;
  auto g_an = apv_an.get(k_an);
  if (g_an != nullptr) {
    k_ci.f0 = person_id;
    auto g_ci = apv_ci.get(k_ci);
    if (g_ci != nullptr) {
      result += g_ci->annotation * g_an->annotation;
    }
  }
}

void DynaGoxCEB3::print() {
  if (print_result) {
    fmt::println(output, "{:.4f}", result);
  }
}